File: 208.implement-trie-prefix-tree.py
#
# @lc app=leetcode id=208 lang=python3
#
# [208] Implement Trie (Prefix Tree)
#

# @lc code=start
from collections import deque


class Trie:
    def __init__(self):
        self.tree = {}
        self.end = "#"  # use this character to indicate the end of a word

    def insert(self, word: str) -> None:
        currDict = self.tree
        w = deque(list(word))
        while w:
            letter = w.popleft()
            if letter not in currDict:
                currDict[letter] = {}
            currDict = currDict[letter]
        currDict[self.end] = True

    def search(self, word: str) -> bool:
        # recursive traverse through the tree along the path built by letters
        # if in the end, we reach "#", return True else False
        w = deque(list(word))
        currDict = self.tree
        while w:
            letter = w.popleft()
            if letter not in currDict:
                return False
            currDict = currDict[letter]
        return self.end in currDict

    def startsWith(self, prefix: str) -> bool:
        # recursive traverse through the tree along the path built by letters
        # if in the end, we reach "#", return True else False
        w = deque(list(prefix))
        currDict = self.tree
        while w:
            letter = w.popleft()
            if letter not in currDict:
                return False
            currDict = currDict[letter]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
# @lc code=end

File: 208.implement-trie-prefix-tree.go
/*
 * @lc app=leetcode id=208 lang=golang
 *
 * [208] Implement Trie (Prefix Tree)
 */

// @lc code=start
package main

var endSymbol rune = 'X'

type Trie struct {
	_t map[rune]Trie
}

func Constructor() Trie {
	t := make(map[rune]Trie)
	trie := Trie{_t: t}
	return trie
}

func (this *Trie) Insert(word string) {
	if len(word) == 0 {
		this._t[endSymbol] = Constructor()
		return
	}
	letter := rune(word[0])
	rest := word[1:]
	if inner, ok := this._t[letter]; !ok {
		newTrie := Constructor()
		this._t[letter] = newTrie
		newTrie.Insert(rest)
	} else {
		inner.Insert(rest)
	}
}

func (this *Trie) Search(word string) bool {
	if len(word) == 0 {
		_, ok := this._t[endSymbol]
		return ok
	}
	letter := rune(word[0])
	rest := word[1:]
	innerTrie, ok := this._t[letter]
	if !ok {
		return false
	}
	return innerTrie.Search(rest)

}

func (this *Trie) StartsWith(prefix string) bool {
	if len(prefix) == 0 {
		return len(this._t) > 0
	}
	letter := rune(prefix[0])
	rest := prefix[1:]
	innerTrie, ok := this._t[letter]
	if !ok {
		return false
	}
	return innerTrie.StartsWith(rest)
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */
// @lc code=end
